numpy/testing/utils.py

```diff
@@ -726,14 +725,13 @@ def chk_same_position(x_id, y_id, hasval='nan'):
                 raise AssertionError(msg)
 
         if isnumber(x) and isnumber(y):
-            x_id, y_id = zeros_like(x, dtype=bool_), zeros_like(y, dtype=bool_)
             if equal_nan:
                 x_isnan, y_isnan = isnan(x), isnan(y)
                 # Validate that NaNs are in the same place
                 if any(x_isnan) or any(y_isnan):
                     chk_same_position(x_isnan, y_isnan, hasval='nan')
-                    x_id |= x_isnan
-                    y_id |= y_isnan
+                    x = x[~x_isnan]
+                    y = y[~y_isnan]
```

Comment:
OK, I am lazy now, but do these assert functions check the shape exactly? Because otherwise, I think there may be a problem with broadcasting if x and y are only broadcastable to one another.